### PR TITLE
[core] Reimplemented strerror to use static messages

### LINKED ISSUE #1585: SIGABRT when calling srt_strerror from connect callback function
Broadcast group socket with 2 links (non blocking mode caller/sender), the called peer is not listening yet. connection times out.

Application receives an SRT stream and try to forward to a group listener peer (not yet started). Connect callback is called with errorcode 1001 (connect timeout) an application tries to log an error message using srt_strerror(errorcode,0)  as argument to a %s format. 

Here the gdb backtrace session:
```
(gdb) bt
#0  0x00007ffff6a3d035 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff6a4079b in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007ffff6a7a22e in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#3  0x00007ffff6a84a66 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#4  0x00007ffff779e91a in std::string::reserve(unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff779ec10 in std::string::append(char const*, unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x0000000000451c89 in append (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:1000
#7  operator+= (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:934
#8  CUDTException::getErrorString (this=0x712f80)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/common.cpp:107
#9  0x0000000000451fc9 in CUDTException::getErrorMessage (this=<optimized out>)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/common.cpp:88
#10 0x0000000000456a61 in srt_strerror (code=<optimized out>, err=0)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/srt_c_api.cpp:249
#11 0x0000000000439101 in SrtConn_ConnectCB (opaq=0x727e10, ns=495907739, errorcode=1001,
    peeraddr=0x7fffe802bf70, token=1) at ../../libraries/naga/mxpmod_lib/srtc_mod.c:1750
#12 0x00000000004b8662 in CRendezvousQueue::updateConnStatus (this=0x7fffe80166f0, rst=RST_AGAIN,
    cst=CONN_AGAIN, response=...)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/queue.cpp:989
#13 0x00000000004bcecb in CRcvQueue::worker (param=0x7fffe801d8d0)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/queue.cpp:1262
#14 0x00007ffff74e9e9a in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#15 0x00007ffff6afd2ed in clone () from /lib/x86_64-linux-gnu/libc.so.6
#16 0x0000000000000000 in ?? ()
(gdb) up
#1  0x00007ffff6a4079b in abort () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) up
#2  0x00007ffff6a7a22e in ?? () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) up
#3  0x00007ffff6a84a66 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) up
#4  0x00007ffff779e91a in std::string::reserve(unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) up
#5  0x00007ffff779ec10 in std::string::append(char const*, unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) up
#6  0x0000000000451c89 in append (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:1000
1000            return this->append(__s, traits_type::length(__s));
(gdb) up
#7  operator+= (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:934
934           { return this->append(__s); }
(gdb) up
#8  CUDTException::getErrorString (this=0x712f80)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/common.cpp:107
107                m_strMsg += ": connection time out";
(gdb) down
#7  operator+= (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:934
934           { return this->append(__s); }
(gdb) down
#6  0x0000000000451c89 in append (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:1000
1000            return this->append(__s, traits_type::length(__s));
(gdb) p *this
$1 = {static npos = <optimized out>,
  _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x7fffd0000998 "Connection setup failure: connection time out"}}
(gdb) bt
#0  0x00007ffff6a3d035 in raise () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff6a4079b in abort () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00007ffff6a7a22e in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#3  0x00007ffff6a84a66 in ?? () from /lib/x86_64-linux-gnu/libc.so.6
#4  0x00007ffff779e91a in std::string::reserve(unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff779ec10 in std::string::append(char const*, unsigned long) ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x0000000000451c89 in append (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:1000
#7  operator+= (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:934
#8  CUDTException::getErrorString (this=0x712f80)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/common.cpp:107
#9  0x0000000000451fc9 in CUDTException::getErrorMessage (this=<optimized out>)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/common.cpp:88
#10 0x0000000000456a61 in srt_strerror (code=<optimized out>, err=0)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/srt_c_api.cpp:249
#11 0x0000000000439101 in SrtConn_ConnectCB (opaq=0x727e10, ns=495907739, errorcode=1001,
    peeraddr=0x7fffe802bf70, token=1) at ../../libraries/naga/mxpmod_lib/srtc_mod.c:1750
#12 0x00000000004b8662 in CRendezvousQueue::updateConnStatus (this=0x7fffe80166f0, rst=RST_AGAIN,
    cst=CONN_AGAIN, response=...)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/queue.cpp:989
#13 0x00000000004bcecb in CRcvQueue::worker (param=0x7fffe801d8d0)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/queue.cpp:1262
#14 0x00007ffff74e9e9a in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#15 0x00007ffff6afd2ed in clone () from /lib/x86_64-linux-gnu/libc.so.6
#16 0x0000000000000000 in ?? ()
(gdb) up
#7  operator+= (__s=0x4ecf62 ": connection time out", this=0x712f98)
    at /usr/include/c++/4.6/bits/basic_string.h:934
934           { return this->append(__s); }
(gdb) up
#8  CUDTException::getErrorString (this=0x712f80)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/common.cpp:107
107                m_strMsg += ": connection time out";
(gdb) p *this
$2 = {<std::exception> = {<No data fields>}, m_iMajor = MJ_SETUP, m_iMinor = MN_TIMEOUT,
  m_iErrno = 0, m_strMsg = {static npos = <optimized out>,
    _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x7fffd0000998 "Connection setup failure: connection time out"}}, m_strAPI = {
    static npos = <optimized out>,
    _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x7ffff79ec4d8 ""}}, m_strDebug = {static npos = <optimized out>,
    _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x7ffff79ec4d8 ""}}, static EUNKNOWN = -1, static SUCCESS = 0,
  static ECONNSETUP = 1000, static ENOSERVER = 1001, static ECONNREJ = 1002, static ESOCKFAIL = 1003,
  static ESECFAIL = 1004, static ECONNFAIL = 2000, static ECONNLOST = 2001, static ENOCONN = 2002,
  static ERESOURCE = 3000, static ETHREAD = 3001, static ENOBUF = 3002, static EFILE = 4000,
  static EINVRDOFF = 4001, static ERDPERM = 4002, static EINVWROFF = 4003, static EWRPERM = 4004,
  static EINVOP = 5000, static EBOUNDSOCK = 5001, static ECONNSOCK = 5002, static EINVPARAM = 5003,
  static EINVSOCK = 5004, static EUNBOUNDSOCK = 5005, static ESTREAMILL = 5009,
  static EDGRAMILL = 5010, static ENOLISTEN = 5006, static ERDVNOSERV = 5007,
  static ERDVUNBOUND = 5008, static EINVALMSGAPI = 5009, static EINVALBUFFERAPI = 5010,
  static EDUPLISTEN = 5011, static ELARGEMSG = 5012, static EINVPOLLID = 5013,
  static EASYNCFAIL = 6000, static EASYNCSND = 6001, static EASYNCRCV = 6002, static ETIMEOUT = 6003,
  static ECONGEST = 6004, static EPEERERR = 7000}
(gdb) up
#9  0x0000000000451fc9 in CUDTException::getErrorMessage (this=<optimized out>)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/common.cpp:88
88          return getErrorString().c_str();
(gdb) up
#10 0x0000000000456a61 in srt_strerror (code=<optimized out>, err=0)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/srt_c_api.cpp:249
249         return(e.getErrorMessage());
(gdb) up
#11 0x0000000000439101 in SrtConn_ConnectCB (opaq=0x727e10, ns=495907739, errorcode=1001,
    peeraddr=0x7fffe802bf70, token=1) at ../../libraries/naga/mxpmod_lib/srtc_mod.c:1750
1750        srtcLog(srtc, LOG_INFO, "connectCB(%p,%x,%d,%s,%d) %s\n",
(gdb) up
#12 0x00000000004b8662 in CRendezvousQueue::updateConnStatus (this=0x7fffe80166f0, rst=RST_AGAIN,
    cst=CONN_AGAIN, response=...)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/queue.cpp:989
989                                 i->m_PeerAddr.get(), token);
(gdb) up
#13 0x00000000004bcecb in CRcvQueue::worker (param=0x7fffe801d8d0)
    at /home/jdube/sandbox/makito2_project/components/vendors/haisrt/srt.git/srtcore/queue.cpp:1262
1262            self->m_pRendezvousQueue->updateConnStatus(rst, cst, unit->m_Packet);
(gdb) up
#14 0x00007ffff74e9e9a in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
(gdb) up
#15 0x00007ffff6afd2ed in clone () from /lib/x86_64-linux-gnu/libc.so.6
(gdb) info thread
  Id   Target Id         Frame
  16   Thread 0x7fffe27fc700 (LWP 3093) "App:KbTop" 0x00007ffff74ee0fe in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
  15   Thread 0x7fffe2ffd700 (LWP 3089) "SRT:RcvQ:w5" 0x00007ffff74f122d in recvmsg ()
   from /lib/x86_64-linux-gnu/libpthread.so.0
  14   Thread 0x7fffe37fe700 (LWP 3088) "SRT:SndQ:w5" 0x00007ffff74edd84 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
  13   Thread 0x7fffe3fff700 (LWP 3087) "SRT:RcvQ:w4" 0x00007ffff74f122d in recvmsg ()
   from /lib/x86_64-linux-gnu/libpthread.so.0
  12   Thread 0x7fffecff9700 (LWP 3086) "SRT:SndQ:w4" 0x00007ffff74edd84 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
  11   Thread 0x7fffed7fa700 (LWP 3085) "SRT:RcvQ:w3" 0x00007ffff74f122d in recvmsg ()
   from /lib/x86_64-linux-gnu/libpthread.so.0
  10   Thread 0x7fffedffb700 (LWP 3084) "SRT:SndQ:w3" 0x00007ffff74edd84 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
* 9    Thread 0x7fffee7fc700 (LWP 3083) "SRT:RcvQ:w2" 0x00007ffff6a3d035 in raise ()
   from /lib/x86_64-linux-gnu/libc.so.6
  8    Thread 0x7fffeeffd700 (LWP 3082) "SRT:SndQ:w2" 0x00007ffff74edd84 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
  7    Thread 0x7ffff4de8700 (LWP 3081) "SRT:TsbPd"
    0x00007ffff74edd84 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
  6    Thread 0x7fffef7fe700 (LWP 3080) "SRT:RcvQ:w1" 0x00007ffff74f122d in recvmsg ()
   from /lib/x86_64-linux-gnu/libpthread.so.0
  5    Thread 0x7fffeffff700 (LWP 3079) "SRT:SndQ:w1" 0x00007ffff74edd84 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
  4    Thread 0x7ffff55e9700 (LWP 3078) "App:Rx" 0x00007ffff74ee0fe in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
  3    Thread 0x7ffff5dea700 (LWP 3077) "UEpoll"
    0x00007ffff74ee0fe in pthread_cond_timedwait@@GLIBC_2.3.2 ()
   from /lib/x86_64-linux-gnu/libpthread.so.0
  2    Thread 0x7ffff65eb700 (LWP 3076) "SRT:GC" 0x00007ffff74ee0fe in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0
  1    Thread 0x7ffff7fe3740 (LWP 3073) "mxptool" 0x00007ffff74ee0fe in pthread_cond_timedwait@@GLIBC_2.3.2 () from /lib/x86_64-linux-gnu/libpthread.so.0

```

### PR DESCRIPTION
**Pull request by ethouris**
Fixes #1585 

The procedure to prepare the error message using `std::string` has been removed and replaced by a function that gets the pointer to a static array of characters that archive the message. The arrays with messages are generated from a comprehensive error message definitions using categories as defined originally in UDT.

Note that generation of the major/minor enums is not supported. After a new error is introduced, it should be added to both definitions, the enum in `srt.h` and in the `generate-error-codes.tcl` script, then the script should be run to regenerate `strerror_defs.cpp` file. Also all enum values must be perfectly monotonic, as a jump over empty code or verification of the monotonicness of the values is not implemented in the generator script.

### CODE REVIEW

File: scripts/generate-error-types.tcl [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		READFAIL  "failure in read"
+		SEEKPFAIL "cannot seek write position"
+		WRITEFAIL "failure in write"
+	}
+
+    NOTSUP "Operation not supported" {
+		NONE           ""
+        ISBOUND        "Cannot do this operation on a BOUND socket"
+        ISCONNECTED    "Cannot do this operation on a CONNECTED socket"
+        INVAL          "Bad parameters"
```
Comment (maxsharabayko) [inline scripts/generate-error-types.tcl:112]: There are some alignment issues (mixed tabs and spaces) here and in the file in general.

File: srtcore/strerror_defs.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+const char* strerror_get_message(size_t major, size_t minor)
+{
+	static const char* const undefined = "UNDEFINED ERROR";
+
+	// Extract the major array
+	if (major >= sizeof(strerror_array_major)/sizeof(const char**))
+	{
+		return undefined;
+	}
```
Comment (maxsharabayko) [inline srtcore/strerror_defs.cpp:151]: Tabs are used instead of spaces for alignment.

File: srtcore/common.cpp [hunk 3]
```diff
@@ -83,230 +83,18 @@ m_iMinor(minor)
...
-   }
-
-   // Adding "errno" information
-   if ((MJ_SUCCESS != m_iMajor) && (0 < m_iErrno))
-   {
-      m_strMsg += ": " + SysStrError(m_iErrno);
-   }
-
-   return m_strMsg;
+    return getErrorMessage();
```
Comment (maxsharabayko) [inline srtcore/common.cpp:97]: It seems `CUDTException::getErrorString()` becomes unused.
Consider removing.

**Review by maxsharabayko — CHANGES_REQUESTED** [review]

File: srtcore/common.cpp (hunk 3 above)
Comment (ethouris) [reply srtcore/common.cpp:102]: It is now used in logging in groups; I'd leave it for development purposes.

**Pull request merged**